### PR TITLE
fix(amplify-category-notifications)

### PR DESCRIPTION
**Pull request by UnleashedMind**
*Issue #, if available:*

*Description of changes:*

- check input params in pinpoint helper
- multi-env-manager code refactor

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-notifications/lib/multi-env-manager.js [hunk 1]
```diff
@@ -183,87 +184,92 @@ async function pushChanges(context, pinpointNotificationsMeta) {
...
+          channels: enabledChannels,
+          Name: serviceMeta.output.Name,
+          Id: serviceMeta.output.Id,
+          Region: serviceMeta.output.Region,
+        };
+        break;
+      }
     }
   }
+  writeTeamProviderInfo(pinpointMeta, context);
```
Comment (yuth) [inline packages/amplify-category-notifications/lib/multi-env-manager.js]: What happens if `pinpointMeta` is undefined?

File: packages/amplify-category-notifications/lib/multi-env-manager.js [hunk 2]
```diff
@@ -183,87 +184,92 @@ async function pushChanges(context, pinpointNotificationsMeta) {
...
+  writeBackendConfig(pinpointMeta, context.amplify.pathManager.getBackendConfigFilePath());
```
Comment (yuth) [inline packages/amplify-category-notifications/lib/multi-env-manager.js]: Can we use this method from `amplify-cli` package.

File: packages/amplify-category-notifications/lib/multi-env-manager.js [hunk 3]
```diff
@@ -183,87 +184,92 @@ async function pushChanges(context, pinpointNotificationsMeta) {
...
-                teamProviderInfo[envName].categories[constants.CategoryName] || {};
-      teamProviderInfo[envName].categories[constants.CategoryName][constants.PinpointName] = {
-        Name: pinpointMeta.Name,
-        Id: pinpointMeta.Id,
-        Region: pinpointMeta.Region,
-      };
-      const jsonString = JSON.stringify(teamProviderInfo, null, 4);
-      fs.writeFileSync(teamProviderInfoFilepath, jsonString, 'utf8');
-    }
+function writeTeamProviderInfo(pinpointMeta, context) {
```
Comment (yuth) [inline packages/amplify-category-notifications/lib/multi-env-manager.js]: This should be moved to `amplify-cli`. This package should not know/care about what is the structure and location of teamProvider.

File: packages/amplify-category-notifications/lib/multi-env-manager.js [hunk 4]
```diff
@@ -183,87 +184,92 @@ async function pushChanges(context, pinpointNotificationsMeta) {
...
+    const jsonString = JSON.stringify(teamProviderInfo, null, 4);
+    fs.writeFileSync(teamProviderInfoFilepath, jsonString, 'utf8');
+  }
+}
 
-    const backendConfigFilePath = context.amplify.pathManager.getBackendConfigFilePath();
-    if (fs.existsSync(backendConfigFilePath)) {
-      const backendConfig = JSON.parse(fs.readFileSync(backendConfigFilePath));
-      backendConfig[constants.CategoryName] = backendConfig[constants.CategoryName] || {};
+function writeBackendConfig(pinpointMeta, backendConfigFilePath) {
```
Comment (yuth) [inline packages/amplify-category-notifications/lib/multi-env-manager.js]: Same as above

File: packages/amplify-category-notifications/lib/multi-env-manager.js [hunk 5]
```diff
@@ -183,87 +184,92 @@ async function pushChanges(context, pinpointNotificationsMeta) {
...
-
-  const currentAmplifyMetaFilePath = context.amplify.pathManager.getCurentAmplifyMetaFilePath();
-  const currentAmplifyMeta = JSON.parse(fs.readFileSync(currentAmplifyMetaFilePath));
-  currentAmplifyMeta[constants.CategoryName] = context.exeInfo.amplifyMeta[constants.CategoryName];
-  jsonString = JSON.stringify(currentAmplifyMeta, null, '\t');
-  fs.writeFileSync(currentAmplifyMetaFilePath, jsonString, 'utf8');
-
-  context.amplify.onCategoryOutputsChange(context);
+function writeAmplifyMeta(categoryMeta, amplifyMetaFilePath) {
+  if (fs.existsSync(amplifyMetaFilePath)) {
```
Comment (yuth) [inline packages/amplify-category-notifications/lib/multi-env-manager.js]: Same as above

File: packages/amplify-category-notifications/lib/multi-env-manager.js (hunk 1 above)
Comment (UnleashedMind) [reply packages/amplify-category-notifications/lib/multi-env-manager.js]: check the function definition. 
it'll remove the pinpoint section in the notifications

File: packages/amplify-category-notifications/lib/multi-env-manager.js (hunk 2 above)
Comment (UnleashedMind) [reply packages/amplify-category-notifications/lib/multi-env-manager.js]: no, notifications need to update to both backend and current backend, it's a special workflow that currently no other categories use.

File: packages/amplify-category-notifications/lib/multi-env-manager.js (hunk 3 above)
Comment (UnleashedMind) [reply packages/amplify-category-notifications/lib/multi-env-manager.js]: agreed, when the cli exposes a method that write to a specific environment in the team provider info.

File: packages/amplify-category-notifications/lib/multi-env-manager.js (hunk 2 above)
Comment (yuth) [reply packages/amplify-category-notifications/lib/multi-env-manager.js]: That might be true, but I don't think serialization of file and writing it to disk should happen here

**Pull request merged**